### ISSUE TITLE
Typing for `_restricted_classmethod` (e.g. for `LightningModule.load_from_checkpoint`) has stopped working for mypy 1.11

### ISSUE THREAD

**Issue by maciejzj** [Lightning-AI/pytorch-lightning#20152]
### Bug description

The current release of mypy 1.11 (1.11.0 & 1.11.1) raises an error for methods decorated with `_restricted_classmethod` (e.g. for `LightningModule.load_from_checkpoint`). From what I have found, there used to be a workaround for tricking type checkers into expected behaviour (from `pytorch_lightning/utilities/model_helpers.py`):

```python
# trick static type checkers into thinking it's a @classmethod
# https://github.com/microsoft/pyright/issues/5865
_restricted_classmethod = classmethod if TYPE_CHECKING else _restricted_classmethod_impl
```

but it seems to have stopped working for mypy 1.11. I am not sure whether this problem is on lightning's side (maybe the workaround needs some update) or this is a mypy bug.

### What version are you seeing the problem on?

v2.2, master

### How to reproduce the bug

Create a file with following contents:

```python
from pytorch_lightning.demos.boring_classes import BoringModel

model = BoringModel.load_from_checkpoint("some.ckpt")
```

Run mypy 1.11 (1.11.0 or 1.11.1) on this file:

```
mypy lit_typing.py
```

### Error messages and logs

```
lit_typing.py:3: error: "object" not callable  [operator]
Found 1 error in 1 file (checked 1 source file)
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:               None
	- available:         False
	- version:           None
* Lightning:
	- lightning-utilities: 0.11.5
	- pytorch-lightning: 2.3.3
	- torch:             2.3.1
	- torchmetrics:      1.4.0.post0
	- torchvision:       0.18.1
* Packages:
	- aiohttp:           3.9.5
	- aiohttp-retry:     2.8.3
	- aiosignal:         1.3.1
	- amqp:              5.2.0
	- annotated-types:   0.7.0
	- antlr4-python3-runtime: 4.9.3
	- anyio:             4.4.0
	- appdirs:           1.4.4
	- appnope:           0.1.4
	- argon2-cffi:       23.1.0
	- argon2-cffi-bindings: 21.2.0
	- arrow:             1.3.0
	- astroid:           3.2.3
	- asttokens:         2.4.1
	- async-lru:         2.0.4
	- asyncssh:          2.15.0
	- atpublic:          4.1.0
	- attrs:             23.2.0
	- autopep8:          2.3.1
	- babel:             2.15.0
	- backcall:          0.2.0
	- beautifulsoup4:    4.12.3
	- billiard:          4.2.0
	- black:             24.4.2
	- bleach:            6.1.0
	- blinker:           1.8.2
	- build:             1.2.1
	- catppuccin:        2.3.0
	- celery:            5.4.0
	- certifi:           2024.7.4
	- cffi:              1.16.0
	- charset-normalizer: 3.3.2
	- click:             8.1.7
	- click-didyoumean:  0.3.1
	- click-plugins:     1.1.1
	- click-repl:        0.3.0
	- colorama:          0.4.6
	- comm:              0.2.2
	- configobj:         5.0.8
	- contourpy:         1.2.1
	- cryptography:      42.0.8
	- cycler:            0.12.1
	- dash:              2.17.1
	- dash-core-components: 2.0.0
	- dash-html-components: 2.0.0
	- dash-table:        5.0.0
	- debugpy:           1.8.2
	- decorator:         5.1.1
	- defusedxml:        0.7.1
	- dictdiffer:        0.9.0
	- dill:              0.3.8
	- diskcache:         5.6.3
	- distlib:           0.3.8
	- distro:            1.9.0
	- docopt:            0.6.2
	- docstring-parser:  0.16
	- dpath:             2.2.0
	- dulwich:           0.22.1
	- dvc:               3.51.2
	- dvc-data:          3.15.1
	- dvc-http:          2.32.0
	- dvc-objects:       5.1.0
	- dvc-render:        1.0.2
	- dvc-studio-client: 0.21.0
	- dvc-task:          0.4.0
	- einops:            0.8.0
	- entrypoints:       0.4
	- executing:         2.0.1
	- fastjsonschema:    2.20.0
	- filelock:          3.15.4
	- flake8:            7.1.0
	- flask:             3.0.3
	- flatten-dict:      0.4.2
	- flufl.lock:        7.1.1
	- fonttools:         4.53.1
	- fqdn:              1.5.1
	- frozenlist:        1.4.1
	- fsspec:            2024.6.1
	- funcy:             2.0
	- gitdb:             4.0.11
	- gitpython:         3.1.43
	- grandalf:          0.8
	- greenlet:          3.0.3
	- gto:               1.7.1
	- h11:               0.14.0
	- h5py:              3.11.0
	- httpcore:          1.0.5
	- httpx:             0.27.0
	- hydra-core:        1.3.2
	- idna:              3.7
	- imageio:           2.34.2
	- importlib-metadata: 8.0.0
	- importlib-resources: 6.4.0
	- iniconfig:         2.0.0
	- ipdb:              0.13.13
	- ipykernel:         6.29.5
	- ipython:           8.12.3
	- ipywidgets:        8.1.3
	- isoduration:       20.11.0
	- isort:             5.13.2
	- iterative-telemetry: 0.0.8
	- itsdangerous:      2.2.0
	- jedi:              0.19.1
	- jinja2:            3.1.4
	- joblib:            1.4.2
	- json5:             0.9.25
	- jsonargparse:      4.32.0
	- jsonpointer:       3.0.0
	- jsonschema:        4.23.0
	- jsonschema-specifications: 2023.12.1
	- jupyter-client:    8.6.2
	- jupyter-core:      5.7.2
	- jupyter-events:    0.10.0
	- jupyter-lsp:       2.2.5
	- jupyter-server:    2.14.2
	- jupyter-server-mathjax: 0.2.6
	- jupyter-server-terminals: 0.5.3
	- jupyterlab:        4.2.3
	- jupyterlab-pygments: 0.3.0
	- jupyterlab-server: 2.27.2
	- jupyterlab-widgets: 3.0.11
	- kiwisolver:        1.4.5
	- klepto:            0.2.5
	- kombu:             5.3.7
	- lazy-loader:       0.4
	- lightning-utilities: 0.11.5
	- markdown-it-py:    3.0.0
	- markupsafe:        2.1.5
	- matplotlib:        3.9.1
	- matplotlib-inline: 0.1.7
	- mccabe:            0.7.0
	- mdurl:             0.1.2
	- mistune:           3.0.2
	- mpmath:            1.3.0
	- msgpack:           1.0.8
	- multidict:         6.0.5
	- mypy:              1.11.1
	- mypy-extensions:   1.0.0
	- nbclient:          0.10.0
	- nbconvert:         7.16.4
	- nbdime:            4.0.1
	- nbformat:          5.10.4
	- neovim:            0.3.1
	- nest-asyncio:      1.6.0
	- networkx:          3.3
	- notebook-shim:     0.2.4
	- numpy:             2.0.0
	- omegaconf:         2.3.0
	- orjson:            3.10.6
	- overrides:         7.7.0
	- packaging:         24.1
	- pandas:            2.2.2
	- pandocfilters:     1.5.1
	- parso:             0.8.4
	- pathspec:          0.12.1
	- pexpect:           4.9.0
	- pickleshare:       0.7.5
	- pillow:            10.4.0
	- pip:               24.0
	- pip-tools:         7.4.1
	- pipreqs:           0.5.0
	- platformdirs:      3.11.0
	- plotly:            5.22.0
	- pluggy:            1.5.0
	- pox:               0.3.4
	- prometheus-client: 0.20.0
	- prompt-toolkit:    3.0.47
	- psutil:            6.0.0
	- ptyprocess:        0.7.0
	- pure-eval:         0.2.2
	- pycodestyle:       2.12.0
	- pycparser:         2.22
	- pydantic:          2.8.2
	- pydantic-core:     2.20.1
	- pydocstyle:        6.3.0
	- pydot:             2.0.0
	- pyflakes:          3.2.0
	- pygit2:            1.15.1
	- pygments:          2.18.0
	- pygtrie:           2.5.0
	- pylint:            3.2.5
	- pynvim:            0.5.0
	- pyparsing:         3.1.2
	- pyproject-hooks:   1.1.0
	- pytest:            8.2.2
	- pytest-mock:       3.14.0
	- python-dateutil:   2.9.0.post0
	- python-json-logger: 2.0.7
	- pytoolconfig:      1.3.1
	- pytorch-lightning: 2.3.3
	- pytz:              2024.1
	- pyyaml:            6.0.1
	- pyzmq:             26.0.3
	- referencing:       0.35.1
	- rentry:            1.0.1
	- requests:          2.32.3
	- retrying:          1.3.4
	- rfc3339-validator: 0.1.4
	- rfc3986-validator: 0.1.1
	- rich:              13.7.1
	- rope:              1.13.0
	- rpds-py:           0.19.0
	- ruamel.yaml:       0.18.6
	- ruamel.yaml.clib:  0.2.8
	- scikit-image:      0.24.0
	- scikit-learn:      1.5.1
	- scipy:             1.14.0
	- scmrepo:           3.3.6
	- semver:            3.0.2
	- send2trash:        1.8.3
	- setuptools:        70.3.0
	- shellingham:       1.5.4
	- shortuuid:         1.0.13
	- shtab:             1.7.1
	- six:               1.16.0
	- smmap:             5.0.1
	- sniffio:           1.3.1
	- snowballstemmer:   2.2.0
	- soupsieve:         2.5
	- sqltrie:           0.11.0
	- stack-data:        0.6.3
	- sympy:             1.13.0
	- tabulate:          0.9.0
	- tenacity:          8.5.0
	- terminado:         0.18.1
	- threadpoolctl:     3.5.0
	- tifffile:          2024.7.2
	- tinycss2:          1.3.0
	- tomlkit:           0.13.0
	- torch:             2.3.1
	- torchmetrics:      1.4.0.post0
	- torchvision:       0.18.1
	- tornado:           6.4.1
	- tqdm:              4.66.4
	- traitlets:         5.14.3
	- typer:             0.12.3
	- types-python-dateutil: 2.9.0.20240316
	- typeshed-client:   2.7.0
	- typing-extensions: 4.12.2
	- tzdata:            2024.1
	- uri-template:      1.3.0
	- urllib3:           2.2.2
	- vine:              5.1.0
	- virtualenv:        20.26.3
	- voluptuous:        0.15.2
	- wcwidth:           0.2.13
	- webcolors:         24.6.0
	- webencodings:      0.5.1
	- websocket-client:  1.8.0
	- werkzeug:          3.0.3
	- wheel:             0.43.0
	- widgetsnbextension: 4.0.11
	- yarg:              0.1.9
	- yarl:              1.9.4
	- zc.lockfile:       3.0.post1
	- zipp:              3.19.2
* System:
	- OS:                Darwin
	- architecture:
		- 64bit
		- 
	- processor:         arm
	- python:            3.12.3
	- release:           23.5.0
	- version:           Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000

</details>

### More info

The problem does not occur for previous mypy versions e.g. mypy 1.10.

**Comment by awaelchli** [comment]
I pushed a fix in b2106ca. The problem was simply that mypy didn't understand the inline if-else statement.
